src/CMakeLists.txt

```diff
@@ -1,87 +1,236 @@
+                COMMAND ${GIT} rev-parse --short HEAD
+                OUTPUT_VARIABLE GIT_VERSION OUTPUT_STRIP_TRAILING_WHITESPACE
+            )
+        endif()
+        SET(BUILD_VERSION ${GIT_VERSION})
+    ENDIF()
+ENDIF()
+message("BUILD_VERSION:${BUILD_VERSION}")
+set(VERSION ${BUILD_VERSION})
+
+# Open qt compile tools
+SET(CMAKE_AUTOUIC ON)
+SET(CMAKE_AUTOMOC ON)
+SET(CMAKE_AUTORCC ON)
+SET(CMAKE_INCLUDE_CURRENT_DIR ON)
+SET(CMAKE_VERBOSE_MAKEFILE ON)
+
+include(CMakePackageConfigHelpers)
+include(GNUInstallDirs)
+include(GenerateExportHeader)
+
+# Record the files that will need to be deleted
+CONFIGURE_FILE(
+    "cmake_uninstall.cmake.in"
+    "${CMAKE_BINARY_DIR}/cmake_uninstall.cmake"
+    IMMEDIATE @ONLY)
+# Create delete target
+ADD_CUSTOM_TARGET(uninstall
+    "${CMAKE_COMMAND}" -P "${CMAKE_BINARY_DIR}/cmake_uninstall.cmake")
 
 SET(BIGINT_DIR ${CMAKE_CURRENT_SOURCE_DIR}/zxing/bigint)
 SET(WIN32_DIR  ${CMAKE_CURRENT_SOURCE_DIR}/zxing/win32/zxing)
 SET(ZXING_DIR  ${CMAKE_CURRENT_SOURCE_DIR}/zxing/zxing)
 
-set(CMAKE_AUTOMOC ON)
-set(CMAKE_WARN_ON)
+SET(BUILD_SHARED_LIBS ON CACHE BOOL "Build shared libs")
 
-set(SOURCES
+if(NOT CMAKE_BUILD_TYPE)
+    set(CMAKE_BUILD_TYPE Release CACHE STRING "Build type" FORCE)
+endif()
 
-    CameraImageWrapper.cpp
-    CameraImageWrapper.h
-    ImageHandler.cpp
-    ImageHandler.h
-    QZXing.cpp
+set(INSTALL_HEADER_FILES
     QZXing.h
     QZXing_global.h
     )
 
-if(QZXING_MULTIMEDIA)
-
-    LIST(APPEND SOURCES QZXingFilter.cpp QZXingFilter.h)
-    add_definitions(-DQZXING_MULTIMEDIA)
+set(HEADER_FILES
+    CameraImageWrapper.h
+    ImageHandler.h
+    ${INSTALL_HEADER_FILES}
+    )
 
-    SET(QZXING_USE_QML ON)
+set(SOURCE_FILES
+    CameraImageWrapper.cpp
+    ImageHandler.cpp
+    QZXing.cpp
+    )
 
-endif(QZXING_MULTIMEDIA)
+add_library(${PROJECT_NAME} ${SOURCE_FILES} ${HEADER_FILES})
+set_target_properties(${PROJECT_NAME} PROPERTIES
+    LIBRARY_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/bin
+    RUNTIME_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/bin
+    ARCHIVE_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/lib
+    PUBLIC_HEADER "${INSTALL_HEADER_FILES}" # Install head files
+    VERSION ${BUILD_VERSION})
+target_compile_definitions(${PROJECT_NAME} PRIVATE
+    QZXING_LIBRARY ZXING_ICONV_CONST)
 
-if(QZXING_USE_QML)
-    LIST(APPEND SOURCES QZXingImageProvider.cpp QZXingImageProvider.h)
-    add_definitions(-DQZXING_QML)
-endif(QZXING_USE_QML)
+# Need qt components
+SET(QT_COMPONENTS Core Gui)
 
-add_library(qzxing "" ${SOURCES})
+option(QZXING_MULTIMEDIA "Use multimedia" ON)
+if(QZXING_MULTIMEDIA)
+    target_sources(${PROJECT_NAME} PRIVATE QZXingFilter.cpp QZXingFilter.h)
+    target_compile_definitions(${PROJECT_NAME} PRIVATE QZXING_MULTIMEDIA)
+    LIST(APPEND QT_COMPONENTS Multimedia)
+    SET(QZXING_QML ON)
+endif(QZXING_MULTIMEDIA)
 
+option(QZXING_QML "Use qml" ${QZXING_QML})
+if(QZXING_QML)
+    target_sources(${PROJECT_NAME} PRIVATE QZXingImageProvider.cpp QZXingImageProvider.h)
```

Comment:
I think you also need to add `target_compile_definitions(${PROJECT_NAME} PRIVATE QZXING_QML)` in here as well, as I'm still having to `add_definitions()` for it to be built.

Reply:
I don't understand why you use this? You only set -DQZXING_QML=ON when run cmake configure.

Reply:
Setting only `QZXING_QML=ON` doesn't work because you are not setting the necessary build define to trigger the feature in the code. Thus, I still have to do `add_definitions(-DQZXING_XML)` in order to get the code compiled properly, otherwise there is API missing from the resulting library.

Reply:
The QZXING_QML is added in follow code:

LIST(APPEND EXPORT_DEFINES QZXING_QML)
# Export defines by configure_file
target_compile_definitions(${PROJECT_NAME} PUBLIC ${EXPORT_DEFINES})

Reply:
commit 4b67917c324f40c4523cd241cb4fb557e6f1c29b